From 48ebbd428514a5a63575dc8fc271979ed9cd5098 Mon Sep 17 00:00:00 2001
From: Kazunori Kobayashi <kkobayas@igel.co.jp>
Date: Mon, 23 Jul 2012 17:14:51 +0900
Subject: [PATCH 11/39] util: revise flush event sequence

Depending on the OpenMax component, OMX_StatePause command must be sent
to the component before OMX_CommandFlush is sent.
After OMX_CommandFlush command is sent, the component state must be
changed to OMX_StateExecuting to resume playback.

Also sending OMX_CommandFlush is required by input and output ports.
The original code has a lack of output ports being flushed.
---
 omx/gstomx_util.c |    7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/omx/gstomx_util.c b/omx/gstomx_util.c
index 0a4b1e7..b17a335 100644
--- a/omx/gstomx_util.c
+++ b/omx/gstomx_util.c
@@ -445,6 +445,7 @@ g_omx_core_wait_for_done (GOmxCore * core)
 void
 g_omx_core_flush_start (GOmxCore * core)
 {
+  g_omx_core_pause (core);
   core_for_each_port (core, g_omx_port_pause);
 }
 
@@ -453,6 +454,9 @@ g_omx_core_flush_stop (GOmxCore * core)
 {
   core_for_each_port (core, g_omx_port_flush);
   core_for_each_port (core, g_omx_port_resume);
+
+  change_state (core, OMX_StateExecuting);
+  wait_for_state (core, OMX_StateExecuting);
 }
 
 /*
@@ -645,6 +649,9 @@ g_omx_port_flush (GOmxPort * port)
       omx_buffer->nFilledLen = 0;
       g_omx_port_release_buffer (port, omx_buffer);
     }
+    OMX_SendCommand (port->core->omx_handle, OMX_CommandFlush, port->port_index,
+        NULL);
+    g_sem_down (port->core->flush_sem);
   } else {
     OMX_SendCommand (port->core->omx_handle, OMX_CommandFlush, port->port_index,
         NULL);
-- 
1.7.9.5

